File: plugins/capacitor-native-websocket/ios/Sources/NativeWebSocketPlugin/NativeWebSocketPlugin.swift
import Capacitor
import Foundation

@objc(NativeWebSocketPlugin)
public class NativeWebSocketPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeWebSocketPlugin"
    public let jsName = "NativeWebSocket"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "connect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "send", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStoredFingerprint", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearStoredFingerprint", returnType: CAPPluginReturnPromise),
    ]

    /// Serial queue protecting all access to `connections` and `lastConnectionId`.
    private let queue = DispatchQueue(label: "com.capacitor.nativewebsocket.pool")
    private var connections: [String: WebSocketManager] = [:]
    private var lastConnectionId: String?

    @objc func connect(_ call: CAPPluginCall) {
        guard let urlString = call.getString("url"),
              let url = URL(string: urlString)
        else {
            call.reject("Missing or invalid 'url' parameter")
            return
        }

        let tls = call.getObject("tls") ?? [:]
        let tlsOpts = TLSOptions(
            required: tls["required"] as? Bool ?? true,
            expectedFingerprint: tls["expectedFingerprint"] as? String,
            allowTOFU: tls["allowTOFU"] as? Bool ?? false,
            storeKey: tls["storeKey"] as? String
        )

        let connectionId = call.getString("connectionId") ?? "__default__"
        let origin = call.getString("origin")

        queue.async { [weak self] in
            guard let self = self else { return }

            // Disconnect any existing connection with the same ID
            self.connections[connectionId]?.disconnect()

            let mgr = WebSocketManager(tlsOptions: tlsOpts)
            self.lastConnectionId = connectionId

            mgr.onOpen = { [weak self] in
                self?.notifyListeners("open", data: ["connectionId": connectionId])
            }

            mgr.onMessage = { [weak self] text in
                self?.notifyListeners("message", data: ["data": text, "connectionId": connectionId])
            }

            mgr.onClose = { [weak self] code, reason in
                self?.queue.async {
                    // Only remove if the manager is still the current one for this ID
                    // (avoids late close from a replaced connection deleting the new one)
                    if self?.connections[connectionId] === mgr {
                        self?.connections.removeValue(forKey: connectionId)
                    }
                }
                var data: [String: Any] = ["code": code, "connectionId": connectionId]
                if let reason { data["reason"] = reason }
                self?.notifyListeners("close", data: data)
            }

            mgr.onError = { [weak self] message in
                self?.notifyListeners("error", data: ["message": message, "connectionId": connectionId])
            }

            mgr.onTLSFingerprint = { [weak self] fingerprint in
                self?.notifyListeners("tlsFingerprint", data: ["fingerprint": fingerprint, "connectionId": connectionId])
            }

            self.connections[connectionId] = mgr
            mgr.connect(url: url, origin: origin)
        }
        call.resolve()
    }

    @objc func send(_ call: CAPPluginCall) {
        guard let data = call.getString("data") else {
            call.reject("Missing 'data' parameter")
            return
        }

        let connectionId = call.getString("connectionId")
        queue.async { [weak self] in
            let cid = connectionId ?? self?.lastConnectionId
            guard let cid = cid, let mgr = self?.connections[cid] else {
                call.reject("WebSocket is not connected")
                return
            }
            mgr.send(data)
            call.resolve()
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        let connectionId = call.getString("connectionId")
        queue.async { [weak self] in
            let cid = connectionId ?? self?.lastConnectionId
            if let cid = cid {
                self?.connections.removeValue(forKey: cid)?.disconnect()
            }
            call.resolve()
        }
    }

    @objc func getStoredFingerprint(_ call: CAPPluginCall) {
        guard let storeKey = call.getString("storeKey") else {
            call.reject("Missing 'storeKey' parameter")
            return
        }

        let fingerprint = TLSCertificateStore.loadFingerprint(storeKey: storeKey)
        call.resolve(["fingerprint": fingerprint as Any])
    }

    @objc func clearStoredFingerprint(_ call: CAPPluginCall) {
        guard let storeKey = call.getString("storeKey") else {
            call.reject("Missing 'storeKey' parameter")
            return
        }

        TLSCertificateStore.clearFingerprint(storeKey: storeKey)
        call.resolve()
    }
}

File: plugins/capacitor-native-websocket/android/src/main/java/com/capacitor/nativewebsocket/NativeWebSocketPlugin.kt
package com.capacitor.nativewebsocket

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@CapacitorPlugin(name = "NativeWebSocket")
class NativeWebSocketPlugin : Plugin() {

    private val connections = ConcurrentHashMap<String, WebSocketManager>()
    /** Tracks the most recently opened connectionId for backward-compat
     *  (send/disconnect without an explicit connectionId). */
    @Volatile private var lastConnectionId: String? = null
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()

    @PluginMethod
    fun connect(call: PluginCall) {
        val url = call.getString("url")
        if (url.isNullOrEmpty()) {
            call.reject("Missing or invalid 'url' parameter")
            return
        }

        // Parse TLS options
        val tlsObj = call.getObject("tls", JSObject())!!
        val tlsOptions = TLSOptions(
            required = tlsObj.optBoolean("required", true),
            expectedFingerprint = tlsObj.optString("expectedFingerprint", null),
            allowTOFU = tlsObj.optBoolean("allowTOFU", false),
            storeKey = tlsObj.optString("storeKey", null),
        )

        val connectionId = call.getString("connectionId") ?: "__default__"
        val origin = call.getString("origin")

        executor.execute {
            // Disconnect any existing connection with the same ID
            connections.remove(connectionId)?.disconnect()

            val mgr = WebSocketManager(context, tlsOptions)
            lastConnectionId = connectionId

            mgr.onOpen = {
                notifyListeners("open", JSObject().apply {
                    put("connectionId", connectionId)
                })
            }

            mgr.onMessage = { text ->
                notifyListeners("message", JSObject().apply {
                    put("data", text)
                    put("connectionId", connectionId)
                })
            }

            mgr.onClose = { code, reason ->
                // Only remove if this manager is still the current one for this ID
                // (avoids late close from a replaced connection deleting the new one)
                connections.remove(connectionId, mgr)
                notifyListeners("close", JSObject().apply {
                    put("code", code)
                    reason?.let { put("reason", it) }
                    put("connectionId", connectionId)
                })
            }

            mgr.onError = { message ->
                notifyListeners("error", JSObject().apply {
                    put("message", message)
                    put("connectionId", connectionId)
                })
            }

            mgr.onTLSFingerprint = { fingerprint ->
                notifyListeners("tlsFingerprint", JSObject().apply {
                    put("fingerprint", fingerprint)
                    put("connectionId", connectionId)
                })
            }

            connections[connectionId] = mgr
            mgr.connect(url, origin)
        }
        call.resolve()
    }

    @PluginMethod
    fun send(call: PluginCall) {
        val data = call.getString("data")
        if (data == null) {
            call.reject("Missing 'data' parameter")
            return
        }

        val connectionId = call.getString("connectionId") ?: lastConnectionId
        val mgr = if (connectionId != null) connections[connectionId] else null
        if (mgr == null) {
            call.reject("WebSocket is not connected")
            return
        }

        mgr.send(data)
        call.resolve()
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        val connectionId = call.getString("connectionId") ?: lastConnectionId
        executor.execute {
            if (connectionId != null) {
                connections.remove(connectionId)?.disconnect()
            }
        }
        call.resolve()
    }

    @PluginMethod
    fun getStoredFingerprint(call: PluginCall) {
        val storeKey = call.getString("storeKey")
        if (storeKey == null) {
            call.reject("Missing 'storeKey' parameter")
            return
        }

        val fingerprint = TLSCertificateStore.loadFingerprint(context, storeKey)
        call.resolve(JSObject().apply {
            put("fingerprint", fingerprint ?: JSObject.NULL)
        })
    }

    @PluginMethod
    fun clearStoredFingerprint(call: PluginCall) {
        val storeKey = call.getString("storeKey")
        if (storeKey == null) {
            call.reject("Missing 'storeKey' parameter")
            return
        }

        TLSCertificateStore.clearFingerprint(context, storeKey)
        call.resolve()
    }
}
